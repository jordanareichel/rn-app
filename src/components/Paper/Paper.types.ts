import {StyleProp, ViewProps, ViewStyle} from 'react-native';

type DefaultProps = {
  size?: number;
  radius?: number;
  center?: boolean;
  circle?: boolean;
  color?: string | string[];
  flexDirection?: 'row' | 'column';
  gradientDirection?: 'vertical' | 'horizontal';
};

export type GradientProps = DefaultProps;

export type PaperProps = {
  gradient?: boolean;
  elevation?: number;
  borderColor?: string;
  borderWidth?: number;
  styleContent?: StyleProp<ViewStyle>;
} & ViewProps &
  DefaultProps;
