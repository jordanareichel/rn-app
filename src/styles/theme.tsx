export const theme = {
  font: {
    sm: '12',
    md: '14',
    lg: '18',
  },
  colors: {
    dark: '#000000',
    light: '#ffffff',
  },
};
