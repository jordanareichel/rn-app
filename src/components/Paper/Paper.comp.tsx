import React from 'react';
import {Wrapper, Content} from './Paper.styles';
import {PaperProps} from './Paper.types';

export const Paper: React.FC<PaperProps> = props => {
  const {
    size,
    color,
    radius,
    center,
    circle,
    gradient,
    children,
    elevation,
    flexDirection,
    styleContent,
    gradientDirection,
    ...rest
  } = props;

  return (
    <Wrapper
      {...Object.assign(
        {size, color, center, circle, radius, flexDirection, elevation},
        rest,
      )}>
      {gradient ? (
        <Content
          colors={[]}
          style={styleContent}
          {...Object.assign({
            size,
            color,
            center,
            circle,
            radius,
            flexDirection,
            gradientDirection,
          })}>
          {children}
        </Content>
      ) : (
        children
      )}
    </Wrapper>
  );
};
