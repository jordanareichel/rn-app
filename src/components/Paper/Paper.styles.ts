import _ from '@utils/object';
import LinearGradient from 'react-native-linear-gradient';
import styled, {css} from 'styled-components/native';
import {PaperProps, GradientProps} from './Paper.types';

const gradientSettings = {
  horizontal: {
    end: {x: 1, y: 0},
    start: {x: 0, y: 0},
  },
};

export const Wrapper = styled.View.attrs<PaperProps>(({elevation = 1}) => ({
  elevation,
  shadowColor: 'black',
  ..._.get(`${elevation - 1}`),
}))<PaperProps>`
  ${({
    size,
    center,
    circle,
    borderWidth = 0,
    flexDirection = 'row',
    borderColor = 'transparent',
    radius = 8,
    color = '#FFFFFF',
  }) => css`
    flex-direction: ${flexDirection};
    border-width: ${borderWidth}px;
    width: ${size ? `${size}px` : 'auto'};
    height: ${size ? `${size}px` : 'auto'};
    align-items: ${center ? 'center' : 'flex-start'};
    justify-content: ${center ? 'center' : 'flex-start'};
    border-color: ${borderColor};
    border-radius: ${circle && size ? Math.round(size / 2) : radius}px;
    background-color: ${typeof color === 'string' ? color : 'transparent'};
  `}
  overflow: hidden;
  position: relative;
`;

export const Content = styled(LinearGradient).attrs<GradientProps>(
  ({gradientDirection = 'vertical', color = '#FFFFFF'}) => {
    const colors = [`${color}_gradient` || color] || color;
    return {
      ...gradientSettings[gradientDirection],
      colors: Array.isArray(colors) ? colors : [colors, colors],
    };
  },
)<GradientProps>`
  ${({size, center, circle, flexDirection = 'row', radius = 8}) => css`
    flex-direction: ${flexDirection};
    align-items: ${center ? 'center' : 'flex-start'};
    justify-content: ${center ? 'center' : 'flex-start'};
    border-radius: ${circle && size ? Math.round(size / 2) : radius}px;
  `}
  flex: 1;
`;
