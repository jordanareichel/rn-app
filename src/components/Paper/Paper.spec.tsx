import {theme} from '@styles/theme';
import React from 'react';
import {ThemeProvider} from 'styled-components/native';
import {Paper} from './Paper.comp';
import TestRenderer from 'react-test-renderer';

describe('Paper component', () => {
  test('renders correctly', () => {
    const tree = TestRenderer.create(
      <ThemeProvider theme={theme}>
        <Paper />
      </ThemeProvider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
