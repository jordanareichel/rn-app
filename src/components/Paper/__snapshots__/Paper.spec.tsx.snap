// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paper component renders correctly 1`] = `
<View
  elevation={1}
  shadowColor="black"
  style={
    Array [
      Object {
        "alignItems": "flex-start",
        "backgroundColor": "#FFFFFF",
        "borderColor": "transparent",
        "borderRadius": 8,
        "borderWidth": 0,
        "flexDirection": "row",
        "height": "auto",
        "justifyContent": "flex-start",
        "overflow": "hidden",
        "position": "relative",
        "width": "auto",
      },
    ]
  }
/>
`;
